processor/signoztailsampler/evaluator.go

```diff
@@ -0,0 +1,141 @@
+package signoztailsampler
+
+import (
+	"github.com/SigNoz/signoz-otel-collector/processor/signoztailsampler/internal/sampling"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.uber.org/zap"
+)
+
+// default evaluator for policyCfg.
+type defaultEvaluator struct {
+	name string
+
+	// true for top level policies
+	root bool
+
+	priority int
+
+	// in case probalistic sampling to be done when filter matches
+	sampler sampling.PolicyEvaluator
+
+	// can be probabilistic, always_include, always_exclude
+	samplingMethod string
+
+	filterOperator string
+	filters        []sampling.PolicyEvaluator
+
+	// the subpolicy evaluators will be processed in order of priority.
+	// the sub-policies will be evaluated prior to evaluating top-level policy,
+	// if any of subpolicy filters match, the sampling method associated with that
+	// sub-policy will be applied. and no further processing will be performed.
+	subpolicies []sampling.PolicyEvaluator
+	logger      *zap.Logger
+}
+
+func NewDefaultEvaluator(logger *zap.Logger, policyCfg *PolicyCfg) sampling.PolicyEvaluator {
+
+	// condition operator to apply on filters (AND | OR)
+	filterOperator := policyCfg.PolicyFilterCfg.FilterOp
+
+	// list of filter evaluators used to decide if this policy
+	// should be applied
+	filters := prepareFilterEvaluators(logger, policyCfg.PolicyFilterCfg)
+	// todo(amol): need to handle situations with zero filters
+
+	// list of sub-policies evaluators
+	subpolicies := make([]sampling.PolicyEvaluator, len(policyCfg.SubPolicies))
+	for _, subRule := range policyCfg.SubPolicies {
```

Comment:
At the moment, we only intent to support one level of sub-policies.   

The structure of policies and sub-policies is similar and the methods of evaluation too. Hence there is a self reference.

Reply:
It would be good to maintain a separate struct for them; the README says policies will be evaluated in a depth-first manner, and this config allows any level of depth and code is recursing. I would suggest defining a sub-policy struct without self-referencing sub policy. This can also become the regular user's processor, not just the sampling rules from UI.

Reply:
this is now handled. the self-ref is removed.